src/embed/init.ts

```diff
@@ -0,0 +1,76 @@
+/* eslint-disable no-underscore-dangle, camelcase */
+import { DataLayerObserver } from '../observer';
+
+/*
+This is where we initialize the DataLayerObserver from this info:
+
+// If true, output from rules are sent only to `console`
+// Default is false
+window['_dlo_preview'] = false;
+
+// If true, rules will run on page load in addition to when there are changes
+// Default is true
+window['_dlo_readOnLoad'] = false;
+
+// Anything on `window` that starts with `_dlo_rules` is read as a rules array
+window['_dlo_rules'] = [
+// rules can go here
+];
+
+window['_dlo_rulesFromAdTeam'] = [
+// rules can also go here
+];
+
+window['_dlo_rulesFromOpsTeam'] = [
+// rules can also go here
+];
+
+*/
+
+function _dlo_collectRules(): any[] {
+  const results: any[] = [];
+  Object.getOwnPropertyNames(window).forEach((propName) => {
+    if (propName.startsWith('_dlo_rules') === false) return;
+
+    const prop = (window as { [key: string]: any })[propName];
+    if (Array.isArray(prop) === false) {
+      /* eslint-disable no-console */
+      console.warn(`window[${propName}] is not an array so will be ignored`);
+      return;
+    }
+
+    prop.forEach((rule: any) => {
+      results.push(rule);
+    });
+  });
+  return results;
+}
+
+function _dlo_initializeFromWindow() {
+  const win = (window as { [key: string]: any });
+
+  // Read flags
+  let preview = false;
+  if (win._dlo_preview === true) {
+    preview = true;
+  }
+  let readOnLoad = false;
+  if (win._dlo_readOnLoad === true) {
+    readOnLoad = true;
+  }
+
+  // Read rules
+  const rules = _dlo_collectRules();
+  if (rules.length === 0) {
+    /* eslint-disable no-console */
+    console.warn('No rules for the Data Layer Observer');
+  }
+
+  win._dlo_observer = new DataLayerObserver({
```

Comment:
We probably need to also have a conditional check on `win._dlo_observer`.  If a domain team acted independently and included `dlo.js` init would try to recreate a `DataLayerObserver`, right?

We could check `win._dlo_observer` and if set, silently stop processing.  But that means a team who dynamically loaded DLO via a tag manager would not have their rules processed.  So there needs to be some incremental processing done for newly found rules, which means we probably need a list of previously processed  _dlo_rules* expandos.  Within init we could check for `win._dlo_observer` and if set, it means that we should look for new rules and then use `win._dlo_observer.processRule(rule) on the incremental ones.`. If needed, `processRule` could be updated to accept a list as convenience.

Reply:
More thoughts.  Previously the pattern was a singleton.  That presents difficulty if one team wants to preview their contributed rule sets (`previewMode` is a global config setting).  We could create a `DataLayerObserver` object for each team -  allowing true separation of rules and config, but that's probably a good amount of design work to itemize and validate the cases.  We can call out in the documentation if only one `DataLayerConfig` rules them all.

Reply:
I've added a check for duplicate assignment of window._dlo_observer that emits an error and bails so we won't create more than one.
For the sake of clarity and simplicity of deployment, what do you think about using the singleton pattern for embed+snippet and if customers want more complex setups then they can directly use the modules?

Reply:
@TrevorFSmith I like the simple approach.  We expect one DLO and corresponding config.  If you load more than one, we won't break but it's not going to do anything.  We will support rules contributed from other teams.  I think that's where we are currently, right?

Reply:
Yep, the DLO will look for anything on `window` that starts with `_dlo_rule` so different teams can add their own array or rules and it will be picked up.